sleap/nn/data/providers.py

```diff
@@ -198,34 +199,61 @@ def py_fetch_lf(ind):
             raw_image = lf.image
             raw_image_size = np.array(raw_image.shape).astype("int32")
 
+            height, width = raw_image_size[:2]
+
             if self.user_instances_only:
                 insts = lf.user_instances
             else:
                 insts = lf.instances
-            insts = [inst for inst in insts if len(inst) > 0]
-            if self.with_track_only:
-                insts = [inst for inst in insts if inst.track is not None]
-            n_instances = len(insts)
-            n_nodes = len(insts[0].skeleton) if n_instances > 0 else 0
-
-            instances = np.full((n_instances, n_nodes, 2), np.nan, dtype="float32")
-            for i, instance in enumerate(insts):
-                instances[i] = instance.numpy()
-
-            skeleton_inds = np.array(
-                [self.labels.skeletons.index(inst.skeleton) for inst in insts]
-            ).astype("int32")
-            track_inds = np.array(
-                [
-                    self.tracks.index(inst.track) if inst.track is not None else -1
-                    for inst in insts
-                ]
-            ).astype("int32")
+
+            instances = []
+
+            for inst in insts:
+
+                # Filter OOB
+                pts = inst.numpy()
+                pts[pts < 0] = np.NaN
+
+                pts[:, 0][pts[:, 0] > width - 1] = np.NaN
+                pts[:, 1][pts[:, 1] > height - 1] = np.NaN
+
+                instance = Instance.from_numpy(pts, inst.skeleton, inst.track)
```

Comment:
Upon further investigation, `Instance.numpy` returns a copy of the underlying `Instance._points`, so no need to create a new `Instance.from_numpy`
```suggestion
```